File: PrintCraftApi/Models/Order.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace PrintCraftApi.Models;

public class Order
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid? UserId { get; set; }

    // Shipping Address
    [Required] public string FullName { get; set; } = string.Empty;
    [Required] public string AddressLine1 { get; set; } = string.Empty;
    public string? AddressLine2 { get; set; }
    [Required] public string City { get; set; } = string.Empty;
    [Required] public string PostalCode { get; set; } = string.Empty;
    [Required] public string PhoneNumber { get; set; } = string.Empty;

    public string Status { get; set; } = "pending_quote";
    public string OrderType { get; set; } = "quote"; // "quote" or "online"
    public decimal DeliveryPrice { get; set; } = 6.95m;
    public decimal OrderDiscountAmount { get; set; } = 0m;
    public decimal? QuotedPrice { get; set; }
    public string? QuoteMessage { get; set; }
    public string? TrackingCode { get; set; }
    public string? TrackingUrl { get; set; }
    public string? InternalNotes { get; set; }
    public string? CustomerNotes { get; set; }
    public bool IsPaid { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    // The list of items in this order
    public List<OrderItem> Items { get; set; } = new();
    public List<Payment> Payments { get; set; } = new();
    public List<OrderCommunication> Communications { get; set; } = new();
    public List<OrderStatusHistory> StatusHistory { get; set; } = new();

    [NotMapped]
    public decimal SubtotalAmount
        => Items.Sum(i => (decimal)i.Price * (i.Count <= 0 ? 1 : i.Count));

    [NotMapped]
    public decimal DiscountAmount
        => Math.Max(OrderDiscountAmount, 0m);

    [NotMapped]
    public decimal FinalTotalAmount
        => Math.Max(SubtotalAmount + Math.Max(DeliveryPrice, 0m) - DiscountAmount, 0m);
}

File: frontend/src/types/order.ts
export interface Order {
  id: string;
  userId?: string | null;
  status:
    | "pending_quote"
    | "quoted"
    | "expired_quote"
    | "pending_payment"
    | "printing"
    | "completed"
    | "shipped"
    | "sent"
    | "delivered"
    | "paid"
    | "failed"
    | "cancelled";
  orderType: "quote" | "online";
  fullName: string;
  addressLine1: string;
  addressLine2?: string;
  city: string;
  postalCode: string;
  phoneNumber: string;
  deliveryPrice?: number;
  serviceFeePrice?: number;
  orderDiscountAmount?: number;
  subtotalAmount?: number;
  discountAmount?: number;
  finalTotalAmount?: number;
  quotedPrice?: number;
  quoteMessage?: string;
  quoteConfirmedAt?: string;
  quoteExpiresAt?: string;
  trackingCode?: string;
  trackingUrl?: string;
  internalNotes?: string;
  customerNotes?: string;
  notes?: OrderNote[];
  isPaid?: boolean;
  updatedAt?: string;
  createdAt: string;
  items: OrderItem[];
  payments?: PaymentAttempt[];
}

export interface OrderNote {
  id: string;
  orderId?: string;
  content: string;
  visibility: "internal" | "customer";
  createdBy?: string;
  createdAt: string;
}

export interface PaymentAttempt {
  id: string;
  orderId: string;
  provider: string;
  reference: string;
  providerPaymentId?: string;
  currency: string;
  amount: number;
  status: string;
  checkoutUrl?: string;
  method?: string;
  failureReason?: string;
  paidAt?: string;
  canceledAt?: string;
  expiredAt?: string;
  failedAt?: string;
  lastWebhookAt?: string;
  webhookAttemptCount?: number;
  lastWebhookPayloadHash?: string;
  lastWebhookError?: string;
  createdAt: string;
  updatedAt?: string;
}

export interface OrderItem {
  id?: string;
  orderId?: string;
  productId?: string;
  imageUrl: string;
  fileUrl?: string;
  fileName: string;
  files?: QuoteItemFile[];
  attachments?: QuoteItemFile[];
  notes?: string;
  size?: string;
  dimensionX?: number;
  dimensionY?: number;
  dimensionZ?: number;
  material: string;
  color: string;
  price: number;
  count: number;
}

export interface QuoteItemFile {
  url: string;
  name: string;
  kind?: "model" | "image" | "other";
}
